test/NuGet.Core.Tests/NuGet.Packaging.Test/PackageExtractorTests.cs

```diff
@@ -3808,6 +3808,51 @@ public async Task InstallFromSourceAsync_LogsSourceOnVerboseLevel()
             }
         }
 
+        [Fact]
+        public async Task InstallFromSourceAsync_WriteSourceToNupkgMetadata()
+        {
+            // Arrange
+            using (var testDirectory = TestDirectory.Create())
+            {
+                var source = Path.Combine(testDirectory, "source");
+                Directory.CreateDirectory(source);
+                var resolver = new VersionFolderPathResolver(Path.Combine(testDirectory, "gpf"));
+                var identity = new PackageIdentity("A", new NuGetVersion("1.2.3"));
+
+                var packageFileInfo = await TestPackagesCore.GeneratePackageAsync(
+                   source,
+                   identity.Id,
+                   identity.Version.ToString(),
+                   DateTimeOffset.UtcNow.LocalDateTime,
+                   "content/A.nupkg");
+
+                using (var packageStream = File.OpenRead(packageFileInfo.FullName))
+                {
+                    var packageExtractionContext = new PackageExtractionContext(
+                        PackageSaveMode.Defaultv3,
+                        XmlDocFileSaveMode.None,
+                        clientPolicyContext: null,
+                        NullLogger.Instance);
+
+                    // Act
+                    await PackageExtractor.InstallFromSourceAsync(
+                        source,
+                        identity,
+                        (stream) => packageStream.CopyToAsync(stream, 4096, CancellationToken.None),
+                        resolver,
+                        packageExtractionContext,
+                        CancellationToken.None);
+
+                    // Assert
+                    var nupkgMetadataPath = resolver.GetNupkgMetadataPath(identity.Id, identity.Version);
+                    Assert.True(File.Exists(nupkgMetadataPath));
+
+                    var nupkgMetadata = NupkgMetadataFileFormat.Read(nupkgMetadataPath);
```

Comment:
Can you add assert it was "Version 2"?
Also maybe hash value assertion too.

Reply:
For this test the version doesn't matter. If later the file version get incremented to 3 or higher, this test has no reason to fail.

Reply:
Do we read back this file sometime later? It seems we only write into it.

Reply:
It's only for manual problem investigation at this time, or for customers who want to implement their own tooling.